swift/apple/FirezoneNetworkExtension/NetworkSettings.swift

```diff
@@ -0,0 +1,195 @@
+          logger.debug("NetworkSettings.apply: Adding IPv6 route: \(address) (prefix length: \(validNetworkPrefixLength))")
+          tunnelIPv6Routes.append(NEIPv6Route(destinationAddress: address, networkPrefixLength: NSNumber(integerLiteral: validNetworkPrefixLength)))
+        }
+      } else {
+        logger.error("NetworkSettings.apply: Ignoring invalid route '\(route, privacy: .public)'")
+      }
+    }
+
+    // We don't really know the connlib gateway IP address at this point, but just using 127.0.0.1 is okay
+    // because the OS doesn't really need this IP address.
+    // NEPacketTunnelNetworkSettings taking in tunnelRemoteAddress is probably a bad abstraction caused by
+    // NEPacketTunnelNetworkSettings inheriting from NETunnelNetworkSettings.
+
+    let tunnelNetworkSettings = NEPacketTunnelNetworkSettings(tunnelRemoteAddress: "127.0.0.1")
+
+    let ipv4Settings = NEIPv4Settings(addresses: [tunnelAddressIPv4], subnetMasks: ["255.255.255.255"])
+    ipv4Settings.includedRoutes = tunnelIPv4Routes
+    tunnelNetworkSettings.ipv4Settings = ipv4Settings
+
+    let ipv6Settings = NEIPv6Settings(addresses: [tunnelAddressIPv6], networkPrefixLengths: [128])
+    ipv6Settings.includedRoutes = tunnelIPv6Routes
+    tunnelNetworkSettings.ipv6Settings = ipv6Settings
+
+    let dnsSettings = NEDNSSettings(servers: [dnsAddress])
+    switch dnsFallbackStrategy {
+      case .systemResolver:
+        // Enable split-DNS. Only those domains matching the resources will be sent to the tunnel's DNS.
+        dnsSettings.matchDomains = resourceDomains
+      case .upstreamResolver:
+        // All DNS queries go to the tunnel's DNS.
+        dnsSettings.matchDomains = [""]
+    }
+    tunnelNetworkSettings.dnsSettings = dnsSettings
+
+    self.hasUnappliedChanges = false
+    packetTunnelProvider.setTunnelNetworkSettings(tunnelNetworkSettings) { error in
+      if let error = error {
+        logger.error("NetworkSettings.apply: Error: \(error, privacy: .public)")
+      } else {
+        guard !self.hasUnappliedChanges else {
+          // Changes were made while the packetTunnelProvider was setting the network settings
+          logger.debug("NetworkSettings.apply: Applying changes made to network settings while we were applying the network settings")
+          self.apply(on: packetTunnelProvider, logger: logger, completionHandler: completionHandler)
+          return
+        }
+        logger.debug("NetworkSettings.apply: Applied successfully")
+      }
+      completionHandler?(error)
+    }
+  }
+}
+
+private extension NetworkSettings {
+  private static func validNetworkPrefixLength(fromString string: String, maximumValue: Int) -> Int {
+    guard let networkPrefixLength = Int(string) else { return 0 }
+    if networkPrefixLength < 0 { return 0 }
+    if networkPrefixLength > maximumValue { return maximumValue }
+    return networkPrefixLength
+  }
+
+  private static func ipv4SubnetMask(networkPrefixLength: Int) -> String {
```

Comment:
As I remember, to use bit math here you want to convert the address to a number first and then use netmask to zero X bytes by shifting them back and forth. We have a code in Panel that does it for reference: https://github.com/firezone/firezone/blob/main/elixir/apps/domain/lib/domain/types/cidr.ex#L40.

Reply:
Had some time on the plane without WiFi so decided to have a bit of fun

```swift
// IPv4
let mask = 0xFFFFFFFF
let maxPrefixLength = 32
let octets = 4

// IPv6
let mask = 0xFFFFFFFF
let maxPrefixLength = 128
let octets = 8

let subnetMask = mask & (mask << (maxPrefixLength - networkPrefixLength)))
var parts: [String] = []
for idx in 0...octets - 1 {
    let part = String(0xFF & (subnetMask >> (idx * 8)), radix: 10)
    parts.append(part)
}

let subnetMaskString = parts.joined(separator: ".")
```

Reply:
Ah, tested with IPv6 and 128-bit ints aren't natively supported, so we'd to repeat with two 64-bit ints or use vBigNum? https://developer.apple.com/documentation/accelerate/veclib/vbignum

Reply:
Using bit math would indeed be more elegant.

I think your code looks great. Except that `for idx in 0...(octets - 1)` should go in reverse rather than forward. Will integrate this into this PR, thanks.

We need this only for IPv4, not IPv6 (only the IPv4 API needs a string-y subnet mask), so we don't need the 128-bit int.

Reply:
Made some minor modifications to your code:
``` swift
  func ipv4SubnetMask(networkPrefixLength: Int) -> String {
    precondition(networkPrefixLength >= 0 && networkPrefixLength <= 32)
    let mask: UInt32 = 0xFFFFFFFF
    let maxPrefixLength = 32
    let octets = 4

    let subnetMask = mask & (mask << (maxPrefixLength - networkPrefixLength))
    var parts: [String] = []
    for idx in 0...(octets - 1) {
      let part = String(UInt32(0x000000FF) & (subnetMask >> ((octets - 1 - idx) * 8)), radix: 10)
      parts.append(part)
    }

    return parts.joined(separator: ".")
  }
}
```